## RSA - The Project

Name:  Daniel Williams

Moodle Email: dawi2291@colorado.edu

Name your file "cuidentikey".ipynb  This means YOUR identikey. 

**Do NOT delete instructions or pre-existing comments in the code. Removal will trigger significat point reductions** 

**Required: Add your own comments as if the instructions and comments are not visible - this allows you to customize your project after the course**



**The Prompt:**

Your boss has been long intrigued by the RSA algorthim and has asked you to create a turorial RSA walk through and exploration project to help the boss and others in the company to understand the basic structures and mathematics of an RSA implementation.

Your project will:
1. Implement the algorithm for breaking codes with a given basic structure. 
2. Explain the steps in an "Inverview Grading Style" with Jupyter Lab Text blocks demostrating and explaining the function as you go. 
3. These demonstrations, explanations, and code comments must demonstrate that you understand the code and what it is doing. 
4. After the basic implementation you will rewrite the code completely with your own improvements, or copy the basic code package and add new features. This will be your custom feature.
5. IMPORTANT - in the first implementation you will keep all instructions, given comments in the code, and use the variable names and set up as given. For your custom feature you may remove these and make the code "your own." ALL CHANGE must be explained in text - why did you change/make improvements? We should be able to see how you adapted the orginal code.






<hr />

# Table of Contents/Grading Structure

#### 0. Bonus +/- Overall structure, code comments, and following directions. (Score may go up or down if overall structure or format is excellent or incomplete). Not extra credit.

## 1. Introduction to your RSA Project (5 points)


## 2. Code Package: FME    (10 points)
        * Convert_Binary_String 
        * FME

## 3. Code Package: Key Generation using Euclidean Algorithms    (10 points)
        * Euclidean Algorithm
        * Extended EA
        * Find_Public_Key_e
        * Find_Private_Key_d

## 4. Code Package: En/Decode Your Messages  (10 points)
        * Convert_Text
        * Convert_Num
        * Encode
        * Decode


## 5. Create a Code Demo(10 points)

## 6. Code Exchange Results  (5 points)

## 7. Write a Narrative (25 points)

## 8. Code Breaking:  (10 points)

## 9. Code Breaking: Complete Examples  (5 points)

## 10. Custom code feature / Advanced options  (10 points) 


    


## 1. Introduction to your RSA Project (5 points) 
<span style="color:DarkRed">Write an introduction/overview here. This should be something someone who has never seen your project could read to understand what is coming. Give a broad summarized overview of RSA components, discuss your custom feature a bit, and  include some preliminary narrative regarding your process. Much of this will be fleshed out in your narrative below, so remember to simply summarize details. This should be a robust paragraph or two of 200 - 300 words.</span>


Welcome to my RSA project. RSA is an encryption scheme that allows secure transfer of information where neither party needs to exchange a "secret" key. To understand the significance of this, consider one approach to sending an encoded message. You want to send a super secret message to your friend, so you encode your message with a super strong encoding algorithm. You send it to your friend. Your friend writes back a message saying, "Thanks for the message, but how do I decode it?" You now have to find a way to communicate to them the "key" to decoding the message. But how can you send them the key securely? If you encrypt the key, we need another key to decrypt it, and so on and so on. This highlights an issue with "secret" key schemes: you need to have a way of securely exchanging keys. This is the "private key" apporach. RSA uses a "public key" approach, which does not rely on the secure transfer of private keys. Rather, both recipient and receiver (and everyone else in the network) have two pairs of keys, a public pair, which everyone knows, and a private pair, which only the owner knows. If you wish to send a message to receipient A, you use their public key to encode a message. This encoding can only be broken by recipient A's private key. Hence, no secret keys are exchanged. The beauty of RSA lies in the fact that there are relatively simple algorithms that can produce such pairs of numbers that satisfy the requirements for being public and private keys. 

This program implements the basic RSA scheme. In addition, I added a custom feature which is an interactive command line program that expands on the basic encoding and decoding of messages. In this program, you choose to generate n-digit keys (though don't go above 7), encrypt and decrypt .txt files, attempt to break keys, as well as encode and decode messages.

## 2. Code Package FME (10 points)

<span style="color:DarkRed"> 

FME is an essential function for RSA to pre-process the messages before the messages are encoded and decoded by the RSA algorithm. 
You have 3 choices of algorithms for this:
* The Rosen Book FME - which will need Convert_Binary_String below.
* Sriram's FME from the video.
* Your own "Square and Mod" function defined with a look up table.
    
*Do not copy code from the internet - it is recognized immediately by our software. In particular, use of bitshifting is forbidden in the first part of the project but may be explored in the custom feature.*
</span>



In [1]:
def Convert_Binary_String(_int):
    """
    Here, you need to define a function that converts an integer to
    a string of its binary expansion. You may or may not need this function. 
    
    For example:
    _int = 345
    bits = 101011001
    """
    '''
    converts an integer to bitstring
    
    _int : int
    
    return string
    '''
    
    l = []
    while _int>0:
        k = _int % 2
        l.insert(0,k)
        _int = _int // 2
    
    ret = ''
    for ch in l:
        ret += str(ch)
        
    return ret

In [2]:
def FME(b, n, m):
    """
    1. Using the fast modular exponentiation algorithm,
    the below function should return b**n mod m.
    As described on page 254. (however, you may input the exponent and 
    then convert it to a string - the book version imports the binary expansion)  
    2. You should use the function defined above Convert_Binary_String() if using the book algorithm.
    3. For this block you MUST use one of the 3 methods above.
    4. Any method using bit-shifting or copied from the internet (even changing varibale names) will result in a 0.
    
    **If you are completely stuck, you may use pow() with a 10pt penalty.**

    You may use the function you developed in your Mastery Workbook - but be sure it is your own
    work, commented, etc. and change inputs as needed.
    """  
    
    '''
    b, n, m : int
    
    returns b ** n mod m using Siriam's fast modular exponentiation algorithm
    
    return int
    '''    
    
    result = 1
    square = b
    while n > 0:
        k = n % 2
        n = int(n/2)
        if k == 1:
            result = result*square % m
        square = square*square % m
    
    return result

## 3. Code Package: Key Generation using Euclidean Algorithms    (10 points)

<span style="color:DarkRed">

The functions below will generate the public and private key pairs which will then be used to create a ciphertext using the public key and then decode the same using the private key. </span>



In [3]:
def Euclidean_Alg(a, b):
    """
    1. Calculate the Greatest Common Divisor of a and b.
    
    2. This version should have only positive inputs and outputs.
    
    3. The function must return a single integer ('x') which is
    the gcd of a and b.
    
    """
    '''
    a, b : positive int
    
    returns gcd(a,b)
    
    return positive int
    '''    
    assert a > 0 and b > 0
    
    # ensure a > b
    if b > a:
        tmp = a
        a = b
        b = tmp
        
    while b > 0:
        k = a % b
        a = b
        b = k
        
    return a
        

In [4]:
def EEA(a, b):
    """
    This is a helper function utilizing Bezout's theorem as discussed in your MW.
    You will follow these same steps closely to construct this function.
    
    This version will return both: 
    1. the GCD of a, b 
    2. Bezout's coefficients in any form you wish. We recommend returning your coefficients as a list or a tuple. 
    HINT: return GCD, (s1, t1)
    
    * Ensure that your inputs are positive integers. Implement these kinds of checks.
    * It might also behoove you to consider reassigning a, b to new coefficients depending on which is greater.
    
    """
    
    '''
    a, b : positive int
    
    returns a list of gcd(a,b) and the Bézout coefficients s1, t1 of a and b respectively 
    
    return [int, int, int]
    '''

    assert a > 0 and b > 0 

    # make sure a > b
    if b > a:
        tmp = a
        a = b
        b = tmp
    
    s1, t1 = 1, 0
    s2, t2 = 0, 1

    while b > 0:
        k = a % b
        q = a // b

        a = b
        b = k

        s1hat, t1hat = s2, t2
        s2hat, t2hat = s1-q*s2, t1-q*t2

        s1, t1 = s1hat, t1hat
        s2, t2 = s2hat, t2hat

    return [a, s1, t1]

In [5]:
import random
import math

def Find_Public_Key_e(p, q):
    """
    Implement this function such that
    it takes 2 primes p and q.
    
    Use the gcd function that you have 
    defined before.
    
    The function should return 2 elements as follows:
    public key: n
    public key: e
    
    HINT: this function will run a loop to find e such 
    that e is relatively prime to (p - 1) (q - 1) 
    and not equal to p or q.
    
    NOTE: There are a number of ways to implement this key feature. 
    You, as the coder, can choose to how to acheive this goal.
    
    """
    '''
    p, q : int
    ASSUME p and q are prime
    
    returns a two-element list [n,e]
    
    returns [int, int]
    '''
    
    n = p*q
    
    modulus = (p-1)*(q-1)
    
    # will choose from list of possible e's
    lst = list(range(2, modulus))

    for i in range(2,modulus):
        e = random.choice(lst)  
        gcd = Euclidean_Alg(e, modulus)
        if gcd == 1 and e != p and e != q:
            break
        else:
            lst.remove(e)
        
    return [n, e]


In [6]:
def Find_Private_Key_d(e, p, q):
    """
    Implement this function to find the decryption exponent d, 
    such that d is the modular inverse of e. 
    
    This will use the Extended Euclidean Algorithm
    
    This function should return the following:
    d: the decryption component.
    
    This is not a single action, and there are multiple methods to create this. 
    
    You may create a helper function or have all code within this function.
    
    Plan ahead before coding this.

    """
    '''
    e, p, q : int
    
    calculates d, the modular inverse of e mod (p-1)*(q-1)
    
    return positive int
    '''

    m = (p-1)*(q-1)
    EEA_result = EEA(e,m)
    d = EEA_result[2]
    
    # ensure return value is non-negative
    while d < 0:
        d += m
    
    return d



## 4. Code Package: En/Decode Your Messages  (10 points)

<span style="color:DarkRed">

1. In this part, you will define two functions `Encode` and `Decode` which will use the public and private keys that you calculated above.
2. Using the public key, the `Encode` function will encode a message and generate the corresponding cipher_text.
3. Using the private key, the `Decode` function will decode a cipher_text and recover the original message.
4. We are proving Convert_Text and Convert_Num for you.</span>



In [7]:
def Convert_Text(_string):
    """
    Define this function such that it takes in a simple 
    string such as "hello" and outputs the corresponding
    standard list of integers (ascii) for each letter in the word hello.
    For example:
    _string = hello
    integer_list = [104, 101, 108, 108, 111]
    
    You may use "ord()"
    """
    '''
    _string : string
    
    converts string to a list of integers using ascii standard encoding
    
    return [int]
    '''
    integer_list = []
    
    for ch in _string:
        integer_list.append(ord(ch))

    return integer_list

In [8]:
def Convert_Num(_list):
    """
    Do the opposite of what you did in the Convert_Text
    function defined above.
    
    Define this function such that it takes in a list of integers
    and outputs the corresponding string (ascii).
    
    For example:
    _list = [104, 101, 108, 108, 111]
    _string = hello
    """
    '''
    _list : [int]
    
    converts a list of integers to their ascii string representations
    
    return string
    '''    

    _string = ''
    for i in _list:
        _string += chr(i)
    return _string

In [9]:
def Encode(n, e, message):
    """
    Here, the message will be a string of characters.
    Use the function Convert_Text from 
    the basic tool set and get a list of numbers.
    
    Encode each of these numbers using n and e and
    return the encoded cipher_text.
    """
    '''
    n, e : int
    message : str
    
    takes a string message and uses a public key to convert it to a list of integers
    
    return [int]
    '''
    cipher_text = []
    
    translated_message = Convert_Text(message)
    
    for ch in translated_message:
        cipher_text.append(FME(ch,e,n))

    return cipher_text

In [10]:
def Decode(n, d, cipher_text):
    """
    Here, the cipher_text will be a list of integers.
    First, you will decrypt each of those integers using 
    n and d.
    
    Later, you will need to use the function Convert_Num from the 
    basic toolset to recover the original message as a string. 
    
    """
    '''
    n, d : int
    cipher_text : [int]
    
    takes a list of integers and uses the private key to decrypt the message
    
    return string
    '''
    decrypted_text = []    

    for number in cipher_text:
        decrypted_text.append(FME(number, d, n))
        
    message = Convert_Num(decrypted_text)

    return message

### <span style="color:white">5.</span>  Create a Code Demo (10 points)
<span style="color:black">

Construct a demonstration of your functions above. This will be a **step-by-step** guide to using your code with a specific example that we can follow using the functions you have created above **using a mix of code and Markdown blocks**. Choose your own "Hello World" type message or a favorite quote of your choosing to code and decode yourself.
 
**This is essentially a test that demonstrates your code with a small walkthrough of how it works. You are simply calling the functions above. You are not using a main function or custom feature here.** Imagine this as a code interview scenario, and your interviewer/future boss wants a preliminary walkthrough of your work. Therefore, some text blocks are essential and helpful for your boss to follow your process. 

Note: You can change and add the type of cell block you are editing with the drop down menu above and go between Code and Markdown.

**This is a mix of code blocks and text blocks** and must discuss all three parts for full credit. You must **show and explain your method and demonstrate (by calling your function as an example)** for generating keys and how your encoding and decoding functions work. 

For full credit: 
    
* When you call FME, explain breifly why we can can't take the "mod" of our number using "%". 
* When you call the EEA, explain why this is so essential and really the key to RSA process.</span>



* Demonstrate and explain how you Generate keys with code and text blocks.

### Code walkthrough ###

In this walkthrough I'll show you how my RSA functions work. We'll see how to find public and private keys, and how to use those keys to encode and decode messages. 

To generate keys, we need to find two primes. This can be done by looking up two primes in a database, such as (https://t5k.org/curios/index.php?start=2&stop=2). Later, in my custom feature section, we will generate them from scratch, but for now, say we visit the website linked above and find two primes, 2707 and 1811. We pass them to our functions which generate public and private keys:

In [11]:
p = 2707
q = 1811

n, e = Find_Public_Key_e(p,q)
d = Find_Private_Key_d(e,p,q)

print(f"This is my public key (n, e): ({n}, {e})")
print(f"This is my private key (n, d): ({n}, {d})")

This is my public key (n, e): (4902377, 2204677)
This is my private key (n, d): (4902377, 3016393)


I'll explain how the two functions work.

In Find_Public_Key_e, the two primes, p and q, are multiplied together. This is n, which is the first element in public key. We then need to find the second element in the public key, which is a number that is relatively prime to (p-1)*(q-1), called e. There are potentially many numbers that satisfy the required property. The code is written to randomly choose numbers from a list of candidates so that different e's can be returned for the same primes, p and q. The following code block demonstrates this. 

In [12]:
for i in range(10):
    print(Find_Public_Key_e(2707, 1811))

[4902377, 1324099]
[4902377, 4306913]
[4902377, 3338801]
[4902377, 233411]
[4902377, 3319711]
[4902377, 1018453]
[4902377, 474169]
[4902377, 3759901]
[4902377, 1185551]
[4902377, 3453437]


Note that the second number changes with each iteration, even though the input values are the same. This allows more than one pair of public and private keys to be generated from the same pair of primes. 

Moving on to Find_Private_Key_d. This function requires as input values the same two primes, p and q, and the second element of the public key, e. The first element of the private key is the same as the public key, so we just need to find the second element of the private key, called d. This number is the modular inverse of e mod (p-1)(q-1). To find the modular inverse, we use the function EEA, which stands for Extended Euclidean Algorithm. 

The EEA is really the key to the whole RSA process. It is an extension of the simpler Euclidean Algorithm, which is a simple algorithm for finding the greatest common divisor of two numbers. In the following code block, we randomly select two integers and find their greatest common divisor (gcd) using the Euclidean Algorithm.

In [13]:
for i in range(10):
    a = random.randint(500,1000)
    b = random.randint(2,500)
    
    gcd = Euclidean_Alg(a, b)
    print(f"The gcd of {a} and {b} is {gcd}.")

The gcd of 794 and 293 is 1.
The gcd of 783 and 493 is 29.
The gcd of 907 and 277 is 1.
The gcd of 825 and 301 is 1.
The gcd of 934 and 304 is 2.
The gcd of 921 and 250 is 1.
The gcd of 725 and 284 is 1.
The gcd of 713 and 192 is 1.
The gcd of 561 and 90 is 3.
The gcd of 863 and 462 is 1.


To find the modular inverse of n mod m, n and m must be relatively prime, which is equivalent to their gcd being 1. Recall that we chose e such that it was relatively prime to (p-1)(q-1). This was done by randomly choosing numbers between 2 and (p-1)(q-1) and calculating their gcd using the Euclidean Algorithm until a suitable e was found.

So we know that e and (p-1)(q-1) are relatively prime. To find the modular inverse, we can work backwards using the steps of the Euclidean algorithm. This, however, requires going through the 'while' loop in the algorithm a second time. A better method exists which is called the Extended Euclidean Algorithm (EEA). The EEA allows us to find a modular inverse without having to go through the entire while loop a second time. The EEA calculates not only the gcd of two integers, but also their Bezout coefficients, which I will explain next.

Say we have two integers, a and b. The Bezout identity states that there exist two integers, i and j, such that ai + bj = gcd(a, b). i and j are called the Bezout coefficients of a and b, respectively. 

The EEA calculates all three of gcd and the Bezout coefficients of any two integers. This is demonstrated below.

In [14]:
gcd, bzt_a, bzt_b = EEA(a, b)

print(f"The gcd of {a} and {b} is {gcd}")
print(f"The Bezout identity is {gcd} = {bzt_a}*{a} + {bzt_b}*{b}")
print(f"{bzt_a}*{a} + {bzt_b}*{b} is {bzt_a*a + bzt_b*b}, which is the same as {gcd}")
print(f"So, the Bezout coefficients of {a} and {b} are {bzt_a} and {bzt_b}, respectively.")

The gcd of 863 and 462 is 1
The Bezout identity is 1 = 53*863 + -99*462
53*863 + -99*462 is 1, which is the same as 1
So, the Bezout coefficients of 863 and 462 are 53 and -99, respectively.


We can see that the the same gcd is found in EEA as in the Euclidean_Alg. 

Why do we need the Bezout coefficients? We need them because we are trying to find the modular inverse of e mod (p-1)(q-1). A modular inverse of a mod m is a number x such that x*a <=> 1 mod m. 

When we calculate the Bezout identity with two coprime numbers a and b, we have 1 = ia + jb. This means that there is a multiple of a that, when added to a multiple of b, equals 1. This is the same as saying ia mod b = 1. So, i is a modular inverse of a mod b. In other words, the EEA allows us to quickly and efficiently find the modular inverse of e mod (p-1)(q-1), which is d, the second element of the private key. This is why the EEA is the key to the whole process. It allows us to efficiently find the private key.

So, we now have our public and private keys. As their names state, the public keys should be made public so that people can send encrypted messages to you. But those messages are only secure if your private key remains secret. Moreover, never share the primes that you used to generate the keys in the first place.

* Demonstrate and explain how you encode a message with code and text blocks.

To encode a message, we use the function ```Encode(n, e, M)```, where M is the message we want to encode and (n, e) is the *recipient's* public key. For this example we will send a message to ourselves. Let's send the message "RSA rocks". Since the recipient is ourself, we use our own public keys (n, e) that we found earlier to generate an encoded message, C: 

In [15]:
M = "RSA rocks"
print("Original message:", M)
C = Encode(n, e, M)
print("Encoded message:", C)

Original message: RSA rocks
Encoded message: [1205323, 1432608, 3191844, 258445, 2622942, 4841643, 2818734, 2984033, 1260198]


And now we have an encoded message. You can see that each integer in the list represents a character in the original message.

I'll explain now how the Encode function works. 

The Encode function uses the public key to convert a message into a ciphertext, which just means that the original message turns into a list of (seemingly) random numbers. 

In the Encode function, we first convert the characters in the message to integers using a standard encoding. We do this using the function Convert_Text, which converts a character to its standard ASCII integer representation: 

In [16]:
converted_message = Convert_Text("hello world")
print(converted_message)

[104, 101, 108, 108, 111, 32, 119, 111, 114, 108, 100]


As you can see, the string "hello world" has been converted into a list of integers, where each character corresponds to an integer.

Next, we use a public key to convert each of these integers into a different integer, which results in an encoded message:

In [17]:
for ch in converted_message:
    print(FME(ch, e, n))

2754603
3103684
1198142
1198142
4841643
258445
1454604
4841643
2622942
1198142
1968975


Here, we have converted the ASCII representations to numbers based on our public key (n, e).

Doing this conversion requires calculating the mod of large numbers. For example, if the public key is (4902377, 1942399), and we need to encode a single character 'h', which is represented by the integer 104 in ASCII, then we need to calculate 104^(1942399) mod 4902377. 104^(1942399) is indeed a large number, but it is not even close to the magnitude of integers used when RSA is properly implemented, which uses numbers that are more than 100 digits long. In these cases, the usual mod function, %, is not fast enough. So instead we use something called Fast Modular Exponentiation, which is called FME in our package. FME is an efficient algorithm that works by using binary arithmetic and properties of exponents to quickly calculate the mod of a**b mod m.

* Demonstrate and explain how you decode the same message with code and text blocks.


To decode a message someone has sent to us, we use the function ```Decode(n, d, C)```, where (n, d) is our private key and C is the encoded message. Let's decode our own message:

In [18]:
print("Encoded message:", C)
M = Decode(n, d, C)
print("Decoded message:", M)

Encoded message: [1205323, 1432608, 3191844, 258445, 2622942, 4841643, 2818734, 2984033, 1260198]
Decoded message: RSA rocks


Our decode function uses the private key to translate the list of integers back into a message. We can see that this is indeed the same message that we sent originally. 

The Decode function works by using FME with the integers in the encoded message and the private key. For example, if we want to decode the first character of our message, we pass the integer along with the private key to the function Decode. Decode works by using FME again to convert the encoded character to the original ASCII integer: 

In [19]:
FME(C[0], d, n)

82

In this example, we convert just the first character of our encoded message using our private key. This returns the integer 82. We then use the function Convert_Num to convert this integer to a character using the standard ASCII conversion:

In [20]:
Convert_Num([82])

'R'

We see that 82 is the ASCII integer for the character 'R', which is indeed the first character in our message "RSA rocks".

We do this to each number in the encoded message, put them together, and we have our original message.

And this is the basic scheme of RSA. 

### 6. Code Exchange Results (5 points) 
<span style="color:white"> Now that your code is working and you are able read and write messages, include 3 complete examples of exchanging code (and exchange is both reading a message and responding) from Piazza here.  Both call your code and output code exchange results here. 
We are just looking to see which 3 codes you have exchanged in Piazza.</span>

In [21]:
# Exchange 1
n,e = 5251, 3        #Public key

n,d = 5251, 3403     #Private key

Message = [2128, 1150, 4250, 1349, 1262, 3336, 2371, 2497, 519, 1262, 1263, 1105, 3336, 1349, 1262, 2310, 1105, 3336, 4115, 762, 2405, 1263, 1105, 3336, 1262, 1349, 1150, 1105, 1262, 506, 1105, 1105, 4723, 2405, 2497, 519, 1262, 1974, 2371, 58, 1262, 519, 1105, 1349, 1262, 4839, 1150, 1105, 2497, 1262, 1974, 2371, 58, 762, 1262, 13, 4679, 1573, 1262, 4115, 2371, 2310, 1105, 1262, 2405, 3336, 1262, 4839, 2371, 762, 1560, 2405, 2497, 519, 3250]

print("Question 1")
print(Decode(n,d,Message))

# Response 1
print("Response:", Encode(n, e, "Dynamite, BTS"))

#-----------------------------------------------#

# Exchange 2
n, e = 9071, 8545             #Public Key

n, d = 9071, 7201             #Private Key

Message = [2615, 861, 5887, 8183, 3506, 5706, 854, 2048, 88, 487, 3506, 6656, 854, 3357, 4161, 4925, 3506, 5911, 854, 3357, 3506, 653, 5887, 8183, 6656, 861, 3506, 854, 2048, 487, 4711, 3506, 5887, 1461, 4925, 3506, 854, 2048, 487, 4711, 3506, 5887, 7617, 5887, 88, 1461, 3506, 5887, 1461, 4925, 3506, 1461, 487, 2048, 487, 4711, 3506, 7617, 487, 8183, 3506, 8570, 88, 6656, 1072, 3506, 854, 8004, 1414, 3506, 313, 8570, 3506, 88, 8183, 3506, 5911, 854, 3357, 4711, 3506, 8004, 5887, 2048, 854, 4711, 88, 8183, 487, 3506, 5706, 854, 2048, 88, 487, 3506, 854, 8004, 3506, 5887, 4161, 4161, 3506, 8183, 88, 5706, 487, 1414]

print()
print("Question 2")
print(Decode(n, d, Message))

# Response 2
print("Response:", Encode(n, e, "Interstellar"))

#--------------------------------------------#

# Exchange 3
n, e = 36437, 35921           #Public Key

n, d = 36437, 28733             #Private Key

Message =[28414, 32539, 30036, 33410, 30036, 32392, 27635, 19942, 32392, 35842, 18874, 30036, 32392, 32662, 27635, 25684, 229, 27635, 35842, 18874, 13845, 24081, 32392, 5916, 24081, 13845, 25684, 30036, 32392, 1210, 35842, 9524, 32392, 34374, 35842, 9524, 24081, 11098, 32392, 24081, 27635, 5082, 30036, 32392, 229, 35842, 32392, 229, 33410, 13845, 24126, 30036, 24081, 32392, 229, 35842, 1974]
print()
print("Question 3")
print(Decode(n, d, Message))

print("Response:", Encode(n, e, "Middle Earth"))

Question 1
What song best describes the feeling you get when your RSA code is working?
Response: [4623, 1974, 2497, 4250, 3283, 2405, 1349, 1105, 1168, 1262, 3942, 4592, 4679]

Question 2
What movie could you watch over and over again and never get sick of? Is it your favorite movie of all time?
Response: [313, 1461, 8183, 487, 4711, 8570, 8183, 487, 4161, 4161, 5887, 4711]

Question 3
Where is one fictional place you would like to travel to?
Response: [25175, 27635, 11098, 11098, 24081, 30036, 32392, 9561, 13845, 33410, 229, 32539]


### 7. Write a Narrative (25 points) 
<span style="color:DarkRed">TWO pages of meaningful narrative about your process - not including the examples -  (800 words) here in this block. Include specifc examples of debugging an challenges referencing your individual process. Responses that are vague, lacking in specific and individual details, or sound like CHatGPT will not recive full credit. Include the following:

* Describe your process. Did you plan ahead? Did you wing it? What have you learned about your own process? 
* Describe the results of exchanging keys and codes with classmates. Did it work the first time or did you need to make adjustments?<span style = "color:white">  Use the word fortitude

* What was most challenging? <span style = "color:white">  Use a cooking metaphor.

* Who helped you with your project?  Which resources were most helpful?

* What was your Best Mistake? (funniest? most frustrating? the one you learned the most from?)</span>

My process for this project was relatively linear. This Jupyter notebook broke down the functions needed for RSA and provided all the necessary documentation to proceed in the coding in a straightforward way. However, in order to understand how all the pieces fit together, I had to go back and watch Prof. Ketelsen's lectures on RSA encryption. I took careful notes during the video to make sure I understood everything. Specifically, the slides about the assumptions in RSA, where the relationships between p, q, n, e, and d, were explained was very helpful in later writing the code. It was also just interesting to see so many different ideas that we had been working with come together -- it was pretty mind blowing to see the Chinese Remainder Theorem tie everything together at the end. This big picture understanding helped me come up with ideas for the custom feature. I learned that, although it is possible to code without understanding the bigger picture, it is much more enjoyable to code when you understand the bigger picture, and it is probably more efficient, since you are thinking about how your function fits in with other functions and you will design it so that it does this well. 

On to the coding part. Siriam's algorithms for FME, EA, and EEA were essential, and saved countless hours. The time I spent the most on were the Find_Keys functions. The first time I coded ```Find_Public_Key_e``` I just used the lowest possible ```e``` I could find. I knew at the time I would change this, but I was focused on getting the whole thing to run. This was a non-essential feature I could add later. After I had the whole thing running, it seemed better to choose testing randomly chosen possible e's so that different ```e```'s could be generated from the same ```(p, q)```. 

For ```Find_Private_Key_d```, ```d``` is the second Bézout coefficient of gcd(e,(p-1)(q-1)), which is calculated using EEA. When I was testing, I discovered that sometimes it returned a negative value. To fix this, I added a ```while``` loop to flip its sign. Understanding modular inverses was essential to understanding why we need the Bezout coefficients. I spent a lot of time in the book/watching the lecture videos understanding modular arithmetic, and I felt like it paid off in this function, which, though only a few lines, really required an understanding of RSA and modular inverses to get right. 

For ```Encode``` and ```Decode```, the first time I coded it I forgot to use FME, so it was just using Python's %. I didn't notice it on my tests because I was using fairly small numbers, but I did catch it when looking through my functions a second time. Another issue I kept running into was the types of the input and outputs. I learned the value of good commenting and later I discovered type annotations which I found essential to keeping everything readable. I also appreciated the ability to hover over functions to peak at their type annotations/comments in VS Code. 

On to exchanging codes. It definitely didn't work the first time. For example, I mistakenly called Encode with the private key instead of the public key. That was "user error" and not a mistake with my code, however. Thankfully, I found that no major changes needed to be made. It was a feeling of great relief when it worked the first time.

The most challenging part of the assignment was the custom feature. Because it is my own thing, I spent much more time with it. I'll talk more about that in the custom feature section. For this beginning part, the most challenging part was just understanding all the pieces and how they fit together. After that, the project was helpfully structured in a way where I could proceed from function to function in a pretty linear manner. But again, actually understanding what was going on required more effort than the actual coding.  

Resources: I believe I used only class resources for this first part. The lecture video on RSA was very helpful in understanding the big picture. Siriam's lectures on the algorithms -- also the proofs! -- were very interesting and of course his going through some of the essential functions for us saved a lot of time. I consulted Rosen a few times at the beginning, but later felt that the other resources provided were more helpful.

I think my "best" mistake was that I wished I did more work on testing how fast my functions were running. I guess in this project I was more focused on getting the coding done and understanding how the pieces fit together. I wish I had spent more time measuring the speed of some of the functions. Since this is my first semester in the program and I don't come from an engineering background, the focus on speed/efficiency is something I'm still getting used to, and I should focus more on measuring such things in the future. On similar lines, I believe I could have finished this project sooner if I had shown more fortitude. In the future, I'd like to complete assignments with more of a buffer before the deadline.

### 8. How to Break Code. (10 points)

<span style="color:DarkRed">


**This section (and section 9 and 10) of the project is only available to students exchanging codes by the Friday (midnight) before the project is due.**

**Implementing a factoring algorithm:**

Begin by coding the basic brute force factorization algorithm given in pseudocode below.

Brute Force Factoring
<pre><code>def factorize(n):
    # n is a number, return the smallest factor of n
    for i from 2 to n-1:
        if i divides n:
            return i
        return FALSE
        </code></pre>

In [22]:
def factorize(n):
    for i in range(2,n):
        if n % i == 0:
            return i
    return False

<span style="color:DarkRed"> After you have tested your RSA package yourself, and tested it with classmates by publishing both the private and public keys on Piazza, post 2 messages with just the public keys for people to break on the "Just Public Keys" thread - use both very small n ‘s (under 1000) for practice. Feel free to come up with n's that create a challenge.
        
* Explain in detail how breaking works in this section using markdown blocks. Demo a short example with a small n. What are the complete steps? What tools do you need? <span style = "color:white">  Use a sports metaphor.
* Try brute forcing a larger value of n, say 15-18 digits long. What do you notice? (If this is taking a while, just ride it out, or Ctrl+C will KeyboardInterrupt the cell block). Provide a sample large n that will take ~3-5 min to crack. 
    * Optional Challenge: Consider illustrating this with python math and time modules. matplotlib and time are good examples. Documentation is available.
* How secure is RSA? Why is it difficult to break codes in RSA? 
* Is our implementation of RSA secure? What are it's flaws? </span>

"Code breaking" here means discovering the private key (n, d) from the public key (n, e). Since n is common to both, we just need to find d. We give ourselves the public key as it is public. Recall that to generate our keys, we used two primes, p and q. So, if we can find these two primes, we can generate all of n, e, and d.

n was calculated by simply multiplying p and q. So we need to reverse this process, which means factoring n. In the brute force algorithm factorize(n) above, we test each integer from 2 to n and see if it divides n. We return the first integer that does so. Since, by the Fundamental Theorem of Arithmetic, every integer greater than 1 has a unique prime factorization, and n was found by mulitplying two primes, the returned integer is guaranteed to be either p or q. The second prime factor is found by dividing n by the returned integer. 

This process is shown in the following code block:

In [23]:
# Choose two primes
p = 2707
q = 1811

n, e = Find_Public_Key_e(p, q)
d = Find_Private_Key_d(e, p, q)
print("The public key is", n, e)
print("The private key is", n, d)

factor1 = factorize(n)
print("One of the factors of n is:", factor1)

factor2 = n//factor1
print("The second factor is:", factor2)

The public key is 4902377 3497363
The private key is 4902377 979067
One of the factors of n is: 1811
The second factor is: 2707


We can see that factor1 and factor2 are indeed the two primes we used to calculate our keys. Now, we simply call our function ```Find_Private_Key_d``` to find ```d```, and use this to decode the message in the same way we did in the code walkthrough:

In [24]:
hacked_d = Find_Private_Key_d(e, factor1, factor2)
print("Private key:", d)
print("Hacked private key:", hacked_d)

Private key: 979067
Hacked private key: 979067


So, is RSA bunk? Far from it. The primes we used to generate our keys were very small -- only 4 digits. In my environment, if I increase their size any higher than 6833 and 6899, so that n = 47485817, my Python kernel dies. This is because factoring is much more computationally intensive than multiplying or modding, and, although there are more efficient algorithms available, even these are not efficient enough to be of practical use.

In [25]:
if False:
    # running this causes my kernel to die
    p = 6899  
    q = 6907
    n,e = Find_Public_Key_e(p,q)
    print(n,e)

    factorize(n)

There are other issues with RSA as we've implemented it which make it vulnerable to frequency analysis. Our messages are encoded as lists of integers, where each integer represents a character in the original message. Since some letters in English are more frequent than others, a code breaker could analyze an intercepted message and correlate the frequency of repeated numbers to the frequency of ascii characters from some dataset. Various hypotheses could be tested which could lead to someone discovering the message's contents without having to factor ```n```.

### 9. Code Breaking: Complete Examples (5 points)

<span style="color:DarkRed">

* Decode one example from Piazza in detail showing all the steps. Use a combination of code and markdown blocks.
* Decode and Respond to 3 "just codes with public keys" in Piazza with different sizes of n. Include complete results here - you may just show the results here.</span>


To break a coded message we use pass n to our factorize function to find one of its prime factors. We then divide n by this factor to find the other prime factor. Now that we have the two prime factors, we can find the private key using ```Find_Private_Key_d``` and decode the message as if we were the intended recipient. 

Using the first post from the "just codes with public keys" thread, we have:

In [26]:
n,e = 703, 113
C = [544, 472, 233, 242, 250, 555, 591, 550, 376, 364, 210, 165, 526, 47, 555, 95, 476, 210, 242, 165, 95, 233, 526, 210, 376, 95, 233, 242, 364, 210, 242, 224, 376, 95, 364, 233, 47, 242, 210, 555, 376, 165, 472, 242, 555, 410, 242, 250, 95, 233, 233, 47, 145]

factor1 = factorize(n)
factor2 = n//factor1

d = Find_Private_Key_d(e,factor1,factor2)

M = Decode(n,d,C)
print("The decoded message is:", M)

The decoded message is: The Conquistador's treasure is buried south of Creed.


In [27]:
# Cailyn Craven posted 
n,e = 130177,13
C = [85308, 48594, 15927, 71285, 61037, 15927, 767, 23406, 71285, 23406, 48594, 12676, 37298, 100459, 71285, 90170, 61037, 38946, 38783, 23406, 71285, 90170, 71285, 113492, 38946, 38946, 86792, 15927, 90170, 37298, 71285, 12676, 767, 71285, 15927, 121493, 15927, 37298, 71285, 12676, 84628, 71285, 29228, 38946, 38783, 71285, 90170, 110238, 15927, 71285, 81798, 110238, 38946, 37298, 100459, 126494, 71285, 29228, 38946, 38783, 71285, 90170, 110238, 15927, 71285, 38946, 37298, 86792, 29228, 71285, 38946, 84628, 84628, 71285, 61037, 29228, 71285, 90170, 71285, 61037, 12676, 23406, 10510]

factor1 = factorize(n)
factor2 = n//factor1

d = Find_Private_Key_d(e,factor1,factor2)

M = Decode(n,d,C)
print("The private key is:", d)
print("The decoded message is:", M)

The private key is: 59749
The decoded message is: The best thing about a Boolean is even if you are wrong, you are only off by a bit.


In [28]:
# Tyler Teichmann posted
n,e = 4440097, 1193
C = [2308405, 79162, 4244727, 3355035, 2035913, 1723627, 3355035, 79162, 2914488, 1723627, 1054412, 4244727, 2035913, 3355035, 2914488, 2035913, 3355035, 1723627, 3134308, 2337190, 765334, 3083536, 1723627, 307890, 2337190, 454005, 2914488, 1723627, 307890, 4244727, 3282085, 1723627, 3159822, 3083536, 2914488, 4244727, 4046808, 1723627, 3355035, 79162, 3220943, 2035913, 1723627, 4046808, 2914488, 3134308, 2164157, 1723627, 3215733, 1723627, 843279, 2337190, 3355035, 1723627, 454005, 2337190, 3502570, 3282085, 1723627, 3355035, 2337190, 1723627, 1867558, 1347399, 3436266, 4358917, 1723627, 2948016, 3220943, 2671253, 2671253, 3220943, 2035913, 2914488, 307890, 2337190, 3282085, 454005, 2035913, 4089773]

factor1 = factorize(n)
factor2 = n//factor1

d = Find_Private_Key_d(e,factor1,factor2)

M = Decode(n,d,C)
print("The private key is:", d)
print("The decoded message is:", M)

The private key is: 936857
The decoded message is: Whats the fastest your code can break this key? I got down to 0.56 milliseconds!


In [29]:
# Adeline Bowman posted
n,e = 623,223
C = [87, 550, 601, 613, 550, 531, 91, 402, 91, 431, 91, 5, 460, 550, 613, 16, 550, 402, 91, 402, 431, 52, 317, 550, 431, 363, 91, 402, 550, 56, 52, 52, 492, 52, 5, 352]

factor1 = factorize(n)
factor2 = n//factor1

d = Find_Private_Key_d(e,factor1,factor2)

M = Decode(n,d,C)
print("The private key is:", d)
print("The decoded message is:", M)

The private key is: 367
The decoded message is: I am visiting my sister this weekend



### 10. Custom CODE Feature  (10 points) ##
<span style="color:DarkRed">- The custom feature should be a "Stand Alone" feature at the end of the notebook. 
- If you need  to use code from above, please copy the entire package and functions again as part of your feature so it can be tested independently. 
- You may remove the give code comments in this section.
- Be sure to provide a demonstration showing that it works. See RSA Guide for more information and suggestions. 
- Include a narrative describing your project.
- All custom features must include a narrative explaining the features and all code must be commented.</span>

**Do not forget the self-grading element.**


In [30]:
import math
import random
import typing

"""
General comments about variable names.

When n, e, and d are used in function signatures, they refer
to these:
    - a public key is an integer tuple (n, e).
    - a private key is an integer tuple (n, d).
"""

def Convert_Binary_String(_int: int) -> str:
    """Converts _int to bitstring"""
    l = []
    while _int > 0:
        k = _int % 2
        l.insert(0,k)
        _int = _int // 2
    
    ret = ''
    for ch in l:
        ret += str(ch)
        
    return ret

def FME(b: int, n: int, m: int) -> int:
    """Returns b ** n mod m
    Fast Modular Exponentiation
    """
    result = 1
    square = b
    while n > 0:
        k = n % 2
        n = int(n/2)
        if k == 1:
            result = result*square % m
        square = square*square % m
    
    return result

def Euclidean_Alg(a: int, b: int) -> int:
    """Returns gcd(a,b)"""
    # ensure a > b
    if b > a:
        tmp = a
        a = b
        b = tmp
        
    while b > 0:
        k = a % b
        a = b
        b = k
        
    return a

def EEA(a: int, b: int) -> tuple[int,int,int]:
    """Returns gcd(a,b) and Bézout coeffs
    Example: EEA(77,43) -> (1,19,-34)
    """
    # ensure a > b
    if b > a:
        tmp = a
        a = b
        b = tmp
    
    s1, t1 = 1, 0
    s2, t2 = 0, 1

    while b > 0:
        k = a % b
        q = a // b

        a = b
        b = k

        s1hat, t1hat = s2, t2
        s2hat, t2hat = s1-q*s2, t1-q*t2

        s1, t1 = s1hat, t1hat
        s2, t2 = s2hat, t2hat

    return (a, s1, t1)

def Find_Public_Key_e(p: int, q: int) -> tuple[int,int]:
    """Calculates public key
    ASSUME p and q are prime
    """
    n = p*q
    
    modulus = (p-1)*(q-1)
    
    # will choose from list of possible e's
    lst = list(range(2,modulus))

    for i in range(2,modulus):
        e = random.choice(lst)  
        gcd = Euclidean_Alg(e, modulus)
        if gcd == 1 and e != p and e != q:
            break
        else:
            lst.remove(e)
            
    return (n, e)

def Find_Private_Key_d(e: int, p: int, q: int) -> int:
    """Calculates d"""
    m = (p-1)*(q-1)
    EEA_result = EEA(e,m)
    d = EEA_result[2]
    
    # in case d is negative
    while d < 0:
        d += m
    
    return d

def Encode(n: int, e: int, message: str) -> list[str]:
    """Encodes message using public key"""
    block_size = find_block_size(n)
    
    cipher_text = []
    
    translated_message = Convert_Text(message)
    

    for ch in translated_message:
        enciphered_message = FME(ch,e,n)
        enciphered_message = str(enciphered_message)
        while len(enciphered_message) < block_size:
            # prepend zeros to have uniform block size
            enciphered_message='0'+enciphered_message
        cipher_text.append(enciphered_message)

    return cipher_text

def Convert_Text(_string: str) -> list[int]:
    """Converts each char in string to ASCII int"""
    integer_list = []
    
    for ch in _string:
        integer_list.append(ord(ch))

    return integer_list

def Convert_Num(_list: list[int]) -> str:
    """Coverts ASCII ints in _list to string"""
    _string = ''
    for i in _list:
        _string += chr(i)
    return _string

def Decode(n: int, d: int, cipher_text: list[int]) -> str:
    """Uses priv key to decode cipher_text"""
    decrypted_text = []    
    for number in cipher_text:
        decrypted_text.append(FME(number, d, n))
    message = Convert_Num(decrypted_text)

    return message

def generate_primes(n: int) -> list[int]:
    """Generates all primes up to n using sieve of eratosthenes"""
    
    # l are odd numbers from 3 to sqrt(n)
    l = list(range(3,math.ceil(math.sqrt(n)),2))
    
    # odd numbers up to n
    i = list(range(3,n,2))
    
    for p in l:
        a = p
        while p < n:
            # generate next multiple of p
            p += a                
            # sieve out p from i
            if p in i:          
                i.remove(p)

    i.insert(0,2)

    return i

def generate_primes_in_range(m: int, n: int) -> list[int]:
    """Generates primes in [m,n]"""
    primes = generate_primes(n)
    primes_in_range = [prime for prime in primes if prime > m]
    return primes_in_range

def find_block_size(n: int) -> int:
    """Determines number digits in n"""
    block_size = 0
    while n > 1:
        n = n/10
        block_size += 1
    return block_size

def encrypt_file(filepath: str, n: int, e: int) -> None:
    """Encrypts .txt file and writes _encrypted.txt to same folder as input file"""
    block_size = find_block_size(n)
    M = []
    with open(filepath, 'r') as file:
        while True:
            block = file.read(block_size)
            if len(block) < block_size:
                block2 = block
                M.append(block2)
            else:
                M.append(block)
            if not block:
                break
    M.pop()
    index = filepath.find('.')
    new_filepath = filepath[:index]
    with open(new_filepath + '_encrypted.txt', 'w') as file:
        for block in M:
            C = Encode(n,e,block)
            for ch in C:
                str_ch = str(ch)
                # prepend 0's to have uniform block size
                while len(str_ch) < block_size:
                    str_ch = '0' + str_ch
                file.write(str_ch)

def decrypt_file(filepath: str, n: int, d: int) -> None:
    """Decrypts a .txt file, writes _decrypted.txt to same folder as input file"""
    block_size = find_block_size(n)
    C = []
    with open(filepath, 'r') as file:
        while True:
            block = file.read(block_size)
            C.append(block)
            if not block:
                break
    C.pop()
    C_int = [int(s) for s in C]
    index = filepath.find('.')
    new_filepath = filepath[:index]
    with open(new_filepath + '_decrypted.txt', 'w') as file:
        file.write(Decode(n,d,C_int))

def factorize(n: int) -> list[int]:
    """Brute force algorithm for factoring n
    returns list of factors, empty list if prime
    """
    factors = []
    for i in range(2,n):
        if n % i == 0:
            factors.append(i)
    return factors

def break_key(n: int, e: int, C: list[int]) -> tuple[int, int]:
    """Given public key and ciphertext, returns private key"""
    factors = factorize(n)
    for i in range(len(factors)):
        p = factors[i]
        for j in range(i + 1,len(factors)):
            q = factors[j]
            d = Find_Private_Key_d(e,p,q)
            print('*'*10)
            print('Decoded message:', Decode(n,d,C))
            print('Private key:', d)
            return n,d

def miller_rabin_test(n: int, k: int) -> str:
    """Returns 'composite' if n is composite, 'probably prime' otherwise
    
    n>2, an odd integer to be tested for primality
    k, the number of rounds of testing to perform
    
    from Wikipedia: https://en.wikipedia.org/wiki/Miller%E2%80%93Rabin_primality_test
    """
    # find s,d such that (2**s)d = n - 1
    m = n-1
    d = m
    s = 0
    while d % 2 == 0:
        d /= 2
        s += 1

    # run test k times; if composite-ness property not found, n is probably prime
    for i in range(k):
        a = random.choice(list(range(2,n-2)))
        x = FME(a,d,n)
        for j in range(s):
            y = x**2 % n
            if y == 1 and x != 1 and x != m:
                return "composite"
            x = y
        if y != 1:
            return "composite"
            
    return "probably prime"

def generate_n_dig_keys(t: int) -> tuple[int,int,int]:
    """Generates primes p,q of approx. len(t) and uses them to produce (n,e,d)"""
    primes = []
    count = 0

    lower = 10**(t-1)+1
    upper = 10**t+1

    for i in range(lower,upper,2):
        n = random.choice(list(range(lower,upper,2)))
        p = miller_rabin_test(n,40)
        if p == "probably prime":
            if p not in primes:
                primes.append(n)
                count += 1
            if count == 2:
                break
                
    p,q = primes
    n = (p-1)*(q-1)
    
    n,e = Find_Public_Key_e(p,q)
    d = Find_Private_Key_d(e,p,q)

    return (n,e,d)

def break_file(n: int, e: int, filepath: str) -> int:
    """Given a public key (n,e) and a filepath, will attempt to break public key"""
    C=[]
    block_size = find_block_size(n)
    with open(filepath, 'r') as file:
        for i in range(10):
            block = file.read(block_size)
            C.append(block)
            
    C_int = [int(x) for x in C]

    d = break_key(n,e,C_int)
    return d

def display_address_book() -> None:
    """Prints fake address book"""
    print("Here is our address book:")
    print('John Doe (264607518881, 127427) \nJane Eyre (290208112387, 459953) \nMoby Duck (136906631761, 257953)\n')

def prompt_pubkey() -> tuple[int,int]:
    """Prompts for public key"""
    prompt = input("Enter public key as xxx, yyy: ")
    n,e = prompt.split(',')
    n,e = int(n),int(e)
    return n,e

def prompt_privkey() -> tuple[int,int]:
    """Prompts for private key"""
    prompt = input("Enter private key as xxx, yyy, or -1 if you do not know: ")
    if prompt == '-1':
        return -1,-1
    n,d = prompt.split(',')
    n,d = int(n),int(d)
    return n,d

def prompt_kind_of_ciphertext(n: int) -> list[int]:
    """Prompts for format of ciphertext
    Returns ciphertext as list of ints
    """
    print('Choose a format to enter your ciphertext:')
    format = input("1=[xxx, yyy, zzz, ...]\n2='xxxyyyzzz...'\n")
    if format == '1':
        return prompt_ciphertext()
    elif format == '2':
        return prompt_ciphertext2(n)
    else:
        return prompt_kind_of_ciphertext()

def prompt_ciphertext() -> list[int]:
    """Prompts for a ciphertext in list form; converts strings to ints
    Returns list of ints
    """
    prompt = input('Enter a ciphertext as [xxx, yyy, zzz, ...]: ')
    prompt = prompt.replace(',','')
    prompt = prompt[1:-1]
    if "'" in prompt:
        prompt = prompt.replace("'", "")
    prompt = prompt.split(' ')
    prompt = [int(c) for c in prompt]
    return prompt

def prompt_ciphertext2(n: int) -> list[int]:
    """Prompts for ciphertext as a single string
    Returns ciphertext as list of ints
    """
    prompt = input('Enter a ciphertext as a single string: ')
    block_size = find_block_size(n)
    num_blocks = len(prompt)//block_size
    C = []
    for i in range(num_blocks):
        C.append(int(prompt[block_size*i:(block_size*i+block_size)]))
    C = [int(c) for c in C]
    return C

def prompt_break_key() -> tuple[int,int] | bool:
    """Prompts whether to break key; 
    if yes, returns private key; if no, false
    """
    print("Do you wish to try to break the key?")
    prompt = prompt_yes_no()
    if prompt == '1':
        n,e = prompt_pubkey()
        C = prompt_kind_of_ciphertext(n)
        n,d = break_key(n,e,C)
        return n,d
    elif prompt == '2':
        return False
    else:
        return prompt_yes_no()

def prompt_yes_no() -> str:
    """Prompts yes or no"""
    prompt = input('1=Yes\n2=No\n')
    if prompt not in ['1','2']:
        return prompt_yes_no()
    else:
        return prompt

def prompt_message() -> str:
    """Prompts for a message to encrypt"""
    prompt = input('Enter a message to encrypt: ')
    return prompt

def prompt_continue() -> bool:
    """Prompts to continue or not
    Returns true if yes, false otherwise
    """
    print(('Continue?'))
    prompt2 = prompt_yes_no()
    if prompt2 == '1':
        return True
    else:
        return False

def prompt_concatenate(C) -> str:
    """Prompts to concatenate or not
    If yes, returns ciphertext as string; False if no
    """
    print('Do you want to concatenate?')
    prompt = prompt_yes_no()
    if prompt == '1':
        S = ''
        for block in C:
            S += str(block)
        return S
    elif prompt == '2':
        return ''
    else:
        return prompt_concatenate(C)

def prompt_digs() -> int:
    """Prompt for length of key"""
    prompt = int(input('Enter approximately how many digits you want your keys to be (must be at least 3 digits): '))
    if prompt < 3:
        return prompt_digs()
    if prompt == 3:
        return 2
    else:
        prompt = prompt//2
        return prompt

def prompt_menu() -> str:
    """Prompt for menu options"""
    prompt = input('\nChoose:\n1=Encrypt Message\n2=Decrypt Message\n3=Encrypt file\n4=Decrypt file\n5=Generate Keys\n6=Quit\n\n')
    return prompt

def main():
    print("\nWelcome to the RSA Encryptor! We support encrypting and decrypting simple messages as well as .txt files.")
    print("You can also generate public and private keys here.")
    
    while True:
        prompt = prompt_menu()
        # Encrypt message option
        if prompt == '1':
            print("Do you have the recipient's public key?")
            prompt2 = prompt_yes_no()
            if prompt2 == '1':
                n,e = prompt_pubkey()
                M = prompt_message()
                C = Encode(n,e,M)
                print('Your encrypted message is below:')
                print(C)
                S = prompt_concatenate(C)
                if S:
                    print('Your encrypted message is below:')
                    print(S)
                if not prompt_continue():
                    break

            elif prompt2 == '2':
                print("Would you like to view our address book?")
                prompt3 = prompt_yes_no()
                if prompt3 == '1':
                    display_address_book()
                    n,e = prompt_pubkey()
                    M = prompt_message()
                    C = Encode(n,e,M)
                    print('Your encrypted message is below:')
                    print(C)
                    S = prompt_concatenate(C)
                    print('Your encrypted message is below:')
                    print(S)
                    if not prompt_continue():
                        break
                else:
                    break
                
        # Decrypt message option
        elif prompt == '2':
            n,d = prompt_privkey()
            if n == -1:
                prompt_break_key()
            else: 
                C = prompt_kind_of_ciphertext(n)
                M = Decode(n,d,C)
                print('Your decrypted message is below:')
                print(M)
            if not prompt_continue():
                break
            
        # Encrypt file option
        elif prompt == '3':
            while True:
                print("Do you have the recipient's public key? ")
                prompt2 = prompt_yes_no()
                if prompt2 == '1':
                    n,e = prompt_pubkey()
                    prompt3 = input("Enter the path to the file you want to encrypt: ")
                    encrypt_file(prompt3,n,e)
                    print(f"\n-->You should now have a file in the same folder as the input file, with '_encrypted' added to the filename.")
                    break
                
                # Does not have pubkey
                elif prompt2 == '2':
                    print("Would you like to view our address book?")
                    prompt3 = prompt_yes_no()
                    if prompt3 == '1':
                        display_address_book()
                        n,e = prompt_pubkey()
                        prompt = input("Enter the path to the file you want to encrypt: ")
                        encrypt_file(prompt,n,e)
                        print(f"\n-->You should now have a file in the same folder as the input file, with '_encrypted' added to the filename.")
                        break
                    else:
                        break
                else:
                    continue
            
        # Decrypt option
        elif prompt == '4':
            prompt2 = input("Enter the path to the file you want to decrypt: ")
            n,d = prompt_privkey()
            if n == -1:
                n,d = prompt_break_key()
            decrypt_file(prompt2, n, d)
            print(f"\n-->You should now have a file in the same folder as the input file, with '_decrypted' added to the filename.")
            if not prompt_continue():
                break
            
        # Generate Keys option
        elif prompt == '5':
            len_digs = prompt_digs()
            n,e,d = generate_n_dig_keys(len_digs)
            print(f"Your public key is {n}, {e}.")
            print(f"Your private key is {n}, {d}.")
            if not prompt_continue():
                break
        elif prompt == '6':
            break
        else:
            continue

    print("Thank you for using the RSA Encryptor!")

if __name__ == "__main__":
    main()



Welcome to the RSA Encryptor! We support encrypting and decrypting simple messages as well as .txt files.
You can also generate public and private keys here.



Choose:
1=Encrypt Message
2=Decrypt Message
3=Encrypt file
4=Decrypt file
5=Generate Keys
6=Quit

 5
Enter approximately how many digits you want your keys to be (must be at least 3 digits):  9


5261 5701
Your public key is 29992961, 12826297.
Your private key is 29992961, 4582633.
Continue?


1=Yes
2=No
 2


Thank you for using the RSA Encryptor!


### Custom Feature Narrative 

There are 5 options in my custom module. We'll start with option 5, which generates keys. The technical problem of generating keys comes down to generating prime numbers. To do this, I implemented the Miller-Rabin primality test. I chose to implement this algorithm because it is widely-used and is not obvious how it works. The test is actually a test of compositeness -- it can only tell us with certainty if a number is composite, and that a number is otherwise "probably prime". To reduce the chances of a false positive, I run the test 40 when finding primes. If the number fails to trigger a "composite" return, then we can be very sure that it is prime. To test the function, I also implemented the Sieve of Eratosthenes, which is the generate_primes(n) function. This function generates all primes up to n. Below, I use it to generate primes up to 10000, and show that there is 100% agreement between it and the miller_rabin_test:

In [40]:
sieved_primes = generate_primes(10000)

# test all odd numbers from 5 to the last prime in sieved_primes
# no output means miller_rabin_test did not produce any false positives
for i in range(5,sieved_primes[-1]+1,2):
    result = miller_rabin_test(i, 40)
    if result == "probably prime" and i not in sieved_primes:
        print(f"Miller-Rabin test mistakenly determined {i} was prime") 

print("finished")

finished


Besides the inherent interest in implementing the Miller-Rabin test, it is also a better way to generate primes within a certain range, which is the functionality I required. To do so with the Sieve of Eratosthenes, you have to first generate all the primes up to the upper bound of your range, and then choose a prime from within that range, which means you've done a good deal of wasted computation finding the primes below the lower bound. For primes of more than several digits this takes far too long to be useable. In my function generate_n_dig_keys(n), I determine the range of numbers I want to search between first, and then randomly test numbers in this range, stopping after I have found two suitable primes.

Once we generate our keys, we can use them to encrypt messages using option 1. The only improvement I made here was to convert each encoded character from an int to a string, and to prepend 0's so that each digit-string is the same size. This was necessary because I wanted to be able to concatenate the blocks together into one long string. If the length of the blocks varies, there is a parsing problem: the decode function can't know ahead of time which blocks are potentially smaller than the others. While this doesn't actually solve the frequency analysis problem, having one long string of digits at least *looks* harder to crack. Block sizes are determined by the size of n, so that the decode function calculates the same block size as the encode function. 

Option 2 -- decypting messages -- follows the same pattern as part 1 of the assignment.

Options 3 and 4 involve encrypting and decypting files. This follows essentially the same pattern as encrypting and decypting messages, except we are reading files. The file reader reads in blocks of strings from the file. Each block is encoded/decoded and appended to a list. This scheme means that whatever the size limit of a list is in python will be the limit of the size of the file that can be encrypted or decrypted.

And that is my custom feature. 


**Please use the grading guide below to self grade your Custom Feature.**

GRADING FOR CUSTOM FEATURE:

10 pts  Wow.  It is amazing.
        Shows initiative and originality. 
        You did something extra special and pushed yourself.
        You went beyond all expectations 
        You broke the rules in a creative way. 
        Your coding and commenting is exceptional.
        Excellent 

7-9 points - It is GOOD! 
        You were a “self starter.” 
         You did everything  requested. 
        All expectations met. 
       	You did a very good job. 
        Good use of commenting.
        Shows mastery of skills.

5-6 points  - Okay.
        Minimum requirements.
        Commenting weak.

0-4 points - It is not finished. 
            Does meet objectives. 
            Did not follow directions. 
            Chose not to do this part (which is a totally fine choice)


## Self grading of custom feature ##

Here in this block, please rate what score out of 8 your custom feature merits given the criteria above:

* For example, if you are brand new to programming, using a main function could be an amazing feature - tell me how this helped you move to a new level.
* In what ways did you push yourself? Did you try something you've never done before?
* Alternatively, if you just did not have the time to do the custom feature, this is actually just fine and a reasonable choice. No judgment.

I would give myself 7/10 for the custom feature. I think it's a decent program, and I did push myself in completing it. However, if I could do it again I would change my priorities before starting. 

Here are some concrete takeaways I got from writing this custom feature:
    - input validation:  takes a lot of time to code. I came up with my own way of  doing it, but I wish I knew about pyinputplus BEFORE starting this. Also, I didn't use try / excepts which I gather is the usual way of doing this. In general, I feel like I don't use try/excepts as much I should. 
    - type annotations: extremely useful for this project, especially because of all the conversions between ints, strs, and lists of strs/ints. 
    - the Python style guide: I didn't read the whole thing but I did read substantial parts, especially related to commenting and type annotations. I reformatted all my functions to (hopefully) conform to the style guide, at least in these respects. 
    - unit testing: while I am good at testing my functions as I go, I want to get better at systematically testing and storing my tests in separate files. I did find myself rewriting tests a lot even though I know this is not best practice. I recently learned about pytest, and plan to use this in future. 
    - priorities: in the end I feel I could've/should've done more with writing/understanding algorithms and testing the speed of things. I'm not sure that I really added much to the original project. Part of what I learned in this was what I should be focusing on as I progress through this program -- writing a cute CL program is nice, but I think my time would have been better spent focusing more on the computer science and math part of things, since that is what this degree is about. I think I missed the opportunity to dive deeper, and for that reason I give myself a 7/10. 